src/main/java/org/apache/maven/plugins/gpg/FilesCollector.java

```diff
@@ -0,0 +1,178 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.MojoFailureException;
+import org.apache.maven.plugin.logging.Log;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.util.FileUtils;
+import org.codehaus.plexus.util.SelectorUtils;
+
+/**
+ * Collects project artifact, the POM, and attached artifacts to be signed.
+ *
+ * @since 3.1.0
+ */
+public class FilesCollector {
+    private final MavenProject project;
+
+    private static final String DEFAULT_EXCLUDES[] =
+            new String[] {"**/*.md5", "**/*.sha1", "**/*.sha256", "**/*.sha512", "**/*.asc"};
+
+    private final String[] excludes;
+
+    private final Log log;
+
+    public FilesCollector(MavenProject project, String[] excludes, Log log) {
+        this.project = project;
+        this.log = log;
+        if (excludes == null || excludes.length == 0) {
+            this.excludes = DEFAULT_EXCLUDES;
+            return;
+        }
+        String newExcludes[] = new String[excludes.length];
+        for (int i = 0; i < excludes.length; i++) {
+            String pattern;
+            pattern = excludes[i].trim().replace('/', File.separatorChar).replace('\\', File.separatorChar);
+            if (pattern.endsWith(File.separator)) {
+                pattern += "**";
+            }
+            newExcludes[i] = pattern;
+        }
+        this.excludes = newExcludes;
+    }
+
+    public List<Item> collect() throws MojoExecutionException, MojoFailureException {
+        List<Item> items = new ArrayList<>();
+
+        if (!"pom".equals(project.getPackaging())) {
+            // ----------------------------------------------------------------------------
+            // Project artifact
+            // ----------------------------------------------------------------------------
+
+            Artifact artifact = project.getArtifact();
+
+            File file = artifact.getFile();
+
+            if (file != null && file.isFile()) {
+                items.add(new Item(file, artifact.getArtifactHandler().getExtension()));
+            } else if (project.getAttachedArtifacts().isEmpty()) {
+                throw new MojoFailureException("The project artifact has not been assembled yet. "
+                        + "Please do not invoke this goal before the lifecycle phase \"package\".");
+            } else {
+                log.debug("Main artifact not assembled, skipping signature generation");
+            }
+        }
+
+        // ----------------------------------------------------------------------------
+        // POM
+        // ----------------------------------------------------------------------------
+
+        File pomToSign =
+                new File(project.getBuild().getDirectory(), project.getBuild().getFinalName() + ".pom");
+
+        try {
+            FileUtils.copyFile(project.getFile(), pomToSign);
```

Comment:
How will this work with producer/consumer?

Reply:
no idea, this is not the purpose of this refactoring :)